### PR TITLE
Migrate multi_margin_loss to ATen (CUDA)

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#61426 Migrate multi_margin_loss to ATen (CUDA)**

Closes gh-24600, closes gh-24601

These operators use custom kernels that aren't well suited to `TensorIterator` style, so this is just changing the CPU code and cleaning up the style.

Differential Revision: [D29648015](https://our.internmc.facebook.com/intern/diff/D29648015)

### CODE REVIEW

File: aten/src/ATen/native/cuda/MultiMarginLoss.cu [hunk 1]
```diff
@@ -0,0 +1,388 @@
...
+              target.data_ptr<int64_t>(),
+              weights.defined() ? weights.data_ptr<scalar_t>() : nullptr,
+              nframe, in_sizes[1],
+              reduction == Reduction::Mean,
+              margin);
+          C10_CUDA_KERNEL_LAUNCH_CHECK();
+        }
+        at::sum_out(out, tmp_output, /*dims=*/IntArrayRef{});
+      }
+    } else {
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/MultiMarginLoss.cu:252]: can this ever happen? input.ndim() <= 2 is checked in the multi_margin_loss_shape_check.

**Review by ngimel — COMMENTED** [review]
This looks good, I have very minor comments

File: aten/src/ATen/native/cuda/MultiMarginLoss.cu (hunk 1 above)
Comment (peterbell10) [reply aten/src/ATen/native/cuda/MultiMarginLoss.cu]: Yeah, I think you're right. I've changed it to an internal assert in the 2d case.

File: aten/src/ATen/native/cuda/MultiMarginLoss.cu [hunk 2]
```diff
@@ -0,0 +1,384 @@
...
+            weights.defined() ? weights.data_ptr<scalar_t>() : nullptr,
+            1,
+            input.dim() < 1 ? input.numel() : input.sizes()[0],
+            reduction == at::Reduction::Mean,
+            margin);
+        C10_CUDA_KERNEL_LAUNCH_CHECK();
+      }
+    } else {
+      auto in_sizes = input.sizes();
+      TORCH_INTERAL_ASSERT(in_sizes.size() == 2);
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/MultiMarginLoss.cu:198]: typo here, hence real build errors.

**Comment by ngimel** [comment]
@ngimel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29648015).

**Pull request closed without merging**